apps/dashboard/src/utils/parseStepVariablesToLiquidVariables.ts

```diff
@@ -5,38 +5,62 @@ export interface LiquidVariable {
   label: string;
 }
 
+export interface ParsedVariables {
+  primitives: LiquidVariable[];
+  arrays: LiquidVariable[];
+  namespaces: LiquidVariable[];
+}
+
 /**
  * Parse JSON Schema and extract variables for Liquid autocompletion.
  * @param schema - The JSON Schema to parse.
- * @returns An array of variable objects suitable for the Liquid language.
+ * @returns An object containing three arrays: primitives, arrays, and namespaces.
  */
-export function parseStepVariablesToLiquidVariables(schema: JSONSchemaDefinition): LiquidVariable[] {
-  const variables: LiquidVariable[] = [];
+export function parseStepVariablesToLiquidVariables(schema: JSONSchemaDefinition): ParsedVariables {
+  const result: ParsedVariables = {
+    primitives: [],
+    arrays: [],
+    namespaces: [],
+  };
 
   function extractProperties(obj: JSONSchemaDefinition, path = ''): void {
-    if (typeof obj === 'boolean') return; // Handle boolean schema
+    if (typeof obj === 'boolean') return;
+
+    if (obj.type === 'object') {
+      // Handle object with additionalProperties
+      if (obj.additionalProperties === true) {
+        result.namespaces.push({
+          type: 'variable',
```

Comment:
Since we split variables by type into three arrays, is the type: 'variable' necessary? If the arrays contain just a string of labels, then it will be much simpler to work with them in the components and we won't need the useMemo.

Reply:
Besides maily, we need to pass them like this in codemirror.